# *Starbucks Rewards Mobile App Offers*

## Introduction

### *Loading Libraries*

In [7]:
# Importing Libraries

import seaborn as sb
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

%matplotlib inline

## Data Wrangling

**Data Wrangling steps:**
- Gather
- Acess
- Clean
    - Define
    - Code
    - Test

### *Gather Data*

In [8]:
# Importing data from json

portfolio_data = pd.read_json(
    "data/portfolio.json", orient='records', lines=True)

profile_data = pd.read_json("data/profile.json", orient='records', lines=True)

transcript_data = pd.read_json(
    "data/transcript.json", orient='records', lines=True)

### *Acess Data*

#### Sample Data

In [15]:
# Glimpse of data
portfolio_data.sample(10)

reward                      channels  difficulty  duration     offer_type  \
6       2  [web, email, mobile, social]          10        10       discount   
5       3  [web, email, mobile, social]           7         7       discount   
2       0          [web, email, mobile]           0         4  informational   
0      10       [email, mobile, social]          10         7           bogo   
3       5          [web, email, mobile]           5         7           bogo   
9       2          [web, email, mobile]          10         7       discount   
4       5                  [web, email]          20        10       discount   
8       5  [web, email, mobile, social]           5         5           bogo   
1      10  [web, email, mobile, social]          10         5           bogo   
7       0       [email, mobile, social]           0         3  informational   

                                 id  
6  fafdcd668e3743c1bb461111dcafc2a4  
5  2298d6c36e964ae4a3e7e9706d1fb8c2  
2  3f207df678b143eea3cee63160fa8bed  
0  ae264e3637204a6fb9bb56bc8210ddfd  
3  9b98b8c7a33c4b65b9aebfe6a799e6d9  
9  2906b810c7d4411798c6938adc9daaa5  
4  0b1e1539f2cc45b7b9fa7c272da2e1d7  
8  f19421c1d4aa40978ebb69ca19b0e20d  
1  4d5c57ea9a6940dd891ad53e9dbe8da0  
7  5a8bc65990b245e5a138643cd4eb9837

In [13]:
profile_data.sample(10)

gender  age                                id  became_member_on   income
4178       M   37  63674022d72c4993a8f6c5edcbf01dcc          20140522  38000.0
11584      F   52  39547839a4a54db1a8c7ecab159182b5          20170121  87000.0
5527       M   44  ec70faa2792f4ee19d67dca4957d285f          20180217  43000.0
15560      M   24  f20859f19f1b4a3987af0b94db1c4cab          20170220  56000.0
1772    None  118  03c7b0b202294fbe8edbd73b0b499fac          20170305      NaN
12777      M   29  68f42565785b4e23aac93af7759bd0dd          20180317  61000.0
12691      F   62  f86af496c16140a5a9e4fbe605a27bb5          20170828  97000.0
16581      M   58  7dae1b05e9ec481c81ccd2186e7ec3d0          20180224  65000.0
6377    None  118  d733b649392d44ac9bf43595a4ea6c83          20180402      NaN
11642      F   40  a182165ed3da4048917d042f7df9ae17          20180201  67000.0

In [14]:
transcript_data.sample(10)

person           event  \
290446  b6008b910a914fe89ccbd9f05154cdda     transaction   
11996   c332c8e9782744c28f0264ac935fb09b  offer received   
120645  08b352c3e09a42e1900892a317f2b84e  offer received   
114923  beb7f9bb1096496fa4a369be8d0c6137  offer received   
116229  6e23a5486c6f424ba3ea13019f6a5806  offer received   
14566   1f99f39237164b17a8b6848f4ce881c3     transaction   
152068  c6715c2bb910495faf267e0f6cb56d6f  offer received   
228163  0537e8df8a5a418ab1fd28c2063ad6f8     transaction   
84129   dcd4fe47f8fd496fb1bbcb673a35d481    offer viewed   
176891  af5f147000ff48199b012df26bf97e15     transaction   

                                                   value  time  
290446                                  {'amount': 3.41}   648  
11996   {'offer id': '5a8bc65990b245e5a138643cd4eb9837'}     0  
120645  {'offer id': '2906b810c7d4411798c6938adc9daaa5'}   336  
114923  {'offer id': '4d5c57ea9a6940dd891ad53e9dbe8da0'}   336  
116229  {'offer id': '2298d6c36e964ae4a3e7e9706d1fb8c2'}   336  
14566                                   {'amount': 10.6}     0  
152068  {'offer id': '4d5c57ea9a6940dd891ad53e9dbe8da0'}   408  
228163                                 {'amount': 12.39}   528  
84129   {'offer id': '2298d6c36e964ae4a3e7e9706d1fb8c2'}   210  
176891                                 {'amount': 21.77}   426

#### Info on the data

In [17]:
# Info of data
portfolio_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   reward      10 non-null     int64 
 1   channels    10 non-null     object
 2   difficulty  10 non-null     int64 
 3   duration    10 non-null     int64 
 4   offer_type  10 non-null     object
 5   id          10 non-null     object
dtypes: int64(3), object(3)
memory usage: 608.0+ bytes


In [18]:
profile_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17000 entries, 0 to 16999
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   gender            14825 non-null  object 
 1   age               17000 non-null  int64  
 2   id                17000 non-null  object 
 3   became_member_on  17000 non-null  int64  
 4   income            14825 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 664.2+ KB


In [20]:
transcript_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 306534 entries, 0 to 306533
Data columns (total 4 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   person  306534 non-null  object
 1   event   306534 non-null  object
 2   value   306534 non-null  object
 3   time    306534 non-null  int64 
dtypes: int64(1), object(3)
memory usage: 9.4+ MB


 #### Description of the data

In [26]:
portfolio_data.describe()

# low standard deviation that means similar reward and difficulty settings on offers.

reward  difficulty   duration
count  10.000000   10.000000  10.000000
mean    4.200000    7.700000   6.500000
std     3.583915    5.831905   2.321398
min     0.000000    0.000000   3.000000
25%     2.000000    5.000000   5.000000
50%     4.000000    8.500000   7.000000
75%     5.000000   10.000000   7.000000
max    10.000000   20.000000  10.000000

In [27]:
profile_data.describe()

# Reasonable standard deviation of age
# High standard deviation in income

age  became_member_on         income
count  17000.000000      1.700000e+04   14825.000000
mean      62.531412      2.016703e+07   65404.991568
std       26.738580      1.167750e+04   21598.299410
min       18.000000      2.013073e+07   30000.000000
25%       45.000000      2.016053e+07   49000.000000
50%       58.000000      2.017080e+07   64000.000000
75%       73.000000      2.017123e+07   80000.000000
max      118.000000      2.018073e+07  120000.000000

In [28]:
transcript_data.describe()

# Nothing to infer

time
count  306534.000000
mean      366.382940
std       200.326314
min         0.000000
25%       186.000000
50%       408.000000
75%       528.000000
max       714.000000

#### Checking for duplicates

In [36]:
# Checking same offer ids which means same offer
portfolio_data.id.duplicated().sum()

# No duplicates in portfolio data

0

In [39]:
# Checking duplicates in profile data on customer id
profile_data.id.duplicated().sum()

# No duplicates

0

In [40]:
# Checking duplicates in transcript data on person(customer id)
transcript_data.person.duplicated().sum()

# These number of people have multiple orders.

289534

- **No duplicates in portfolio data**
- **No duplicates in profile data**
- **289534 number of people have multiple orders. What to do will be decided later depending on data.**

#### Checking for Null values

In [43]:
# Portfolio data
print(portfolio_data.isna().sum())

# No null values

reward        0
channels      0
difficulty    0
duration      0
offer_type    0
id            0
dtype: int64


In [51]:
# Profile data
print(profile_data.isna().sum())

# Gender has null values
# Checking percentage of null values
print('\nPercentage of data missing: ', profile_data.gender.isna().sum() / profile_data.shape[0])

gender              2175
age                    0
id                     0
became_member_on       0
income              2175
dtype: int64

Percentage of data missing:  0.12794117647058822


In [47]:
# Transcript data
transcript_data.isna().sum()

# No null values

person    0
event     0
value     0
time      0
dtype: int64

- **No missing data in portfolio and transcript but about 13% gender data missing in profile and that can be removed.**
- This may increase as we have to merge data of profile and transcript.

### Access Summary

***Tidiness Issues***
- Portfolio data has channels in the form of list
- Transcript data has value column containg items in the form of dictionary.

***Quality Issues***
- Profile data has gender and income column containing null values

### *Clean Data*

In [56]:
# Creating copies of the data
portfolio_data_clean = portfolio_data.copy()
profile_data_clean = profile_data.copy()
transcript_data_clean = transcript_data.copy()

**Define** 

**Code**

**Test**

**Define** 

**Code**

**Test**

**Define** 

**Code**

**Test**